File: p1/java/app/src/main/java/cse262/slang/Slang.java
package cse262.slang;

import java.nio.file.Files;
import java.nio.file.Path;

import cse262.slang.Scanner.Scanner;
import cse262.slang.Scanner.XmlTokenWriter;
import cse262.slang.Scanner.Scanner.ScanError;

/**
 * Slang implements an interpreter for a Scheme-like language
 *
 * NB: This is phase 1, so there's just a scanner
 */
public class Slang {
    /**
     * Read the entire contents of a file as a String. On error, a message
     * will be printed, and an empty string will be returned.
     *
     * @param fileName The name of the file to open
     *
     * @return An empty string (`""`) on any error. Otherwise, the contents of
     *         the file
     */
    private static String getFile(String fileName) {
        try {
            Path path = Path.of(fileName);
            return Files.readString(path);
        } catch (Exception ex) {
            System.err.println("Error: Unable to open " + fileName);
            return "";
        }
    }

    public static void main(String[] args) {
        // Get the command-line arguments. If help is requested or needed,
        // print help and exit immediately.
        var parsedArgs = new Args(args);
        if (parsedArgs.mode == Args.Modes.HELP) {
            Args.printHelp();
            return;
        }

        // Load the source file
        String codeToRun = getFile(parsedArgs.fileName);

        // SCAN mode: read the source code, turn it into tokens, print the
        // tokens as XML.
        if (parsedArgs.mode == Args.Modes.SCAN) {
            try {
                var tokens = new Scanner().scanTokens(codeToRun);
                var writer = new XmlTokenWriter();
                writer.writeXmlToStream(tokens, System.out);
            } catch (ScanError se) {
                System.out.println(se.getMessage());
            }
        }
    }
}

File: p1/python/slang.py
import sys
import slang_scanner
import slang_scanner_xml


def get_file(filename):
    """Read a file and return its contents as a single string.

    Args:
        filename: The path to the file to read.

    Returns:
        The contents of the file, or an empty string if not found.
    """
    try:
        source_file = open(filename, "r")
        code = source_file.read()
        source_file.close()
        return code
    except FileNotFoundError:
        print("Error: file not found", file=sys.stderr)
        return ""


def print_help():
    """Print a help message"""
    print("slang -- An interpreter for a Scheme-like language (Python version)")
    print("  Usage: slang [mode] [filename]")
    print("    * The named file will be loaded and evaluated")
    print("  Modes:")
    print("    -help             Display this message and exit")
    print("    -scan             Turn slang code into XML tokens")


def main(args):
    """Run the Scheme interpreter.

    Args:
        args: The command-line arguments.
    """
    # Parse the command-line arguments.  Make sure exactly one valid mode is
    # given, and (for every mode but -help) exactly one filename
    filename, mode, num_modes, num_files = "", "", 0, 0
    for a in args:
        if a in ["-help", "-scan"]:
            mode = a
            num_modes += 1
        else:
            filename = a
            num_files += 1
    if num_modes != 1 or (mode == "-help" and num_files != 0) or (mode != "-help" and num_files != 1):
        return print_help()

    code_to_run = get_file(filename)

    if mode == "-scan":
        tokens = slang_scanner.scan_tokens(code_to_run)
        print(slang_scanner_xml.to_xml_string(tokens))


# In python, this is how we get main() to run when we invoke this program via
# `python3 slang.py ...`
if __name__ == "__main__":
    main(sys.argv[1:])
